receiver/receivercreator/runner.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receivercreator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/config"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"go.uber.org/zap"
+)
+
+// runner starts and stops receiver instances.
+type runner interface {
+	// start a receiver instance from its static config and discovered config.
+	start(receiver receiverConfig, discoveredConfig userConfigMap) (component.Receiver, error)
+	// shutdown a receiver.
+	shutdown(rcvr component.Receiver) error
+}
+
+// receiverRunner handles starting/stopping of a concrete subreceiver instance.
+type receiverRunner struct {
+	logger       *zap.Logger
+	nextConsumer consumer.MetricsConsumerOld
+	idNamespace  string
+	ctx          context.Context
+	host         component.Host
+}
+
+var _ runner = (*receiverRunner)(nil)
+
+// start a receiver instance from its static config and discovered config.
+func (run *receiverRunner) start(receiver receiverConfig, discoveredConfig userConfigMap) (component.Receiver, error) {
+	factory := run.host.GetFactory(component.KindReceiver, receiver.typeStr)
+
+	if factory == nil {
+		return nil, fmt.Errorf("unable to lookup factory for receiver %q", receiver.typeStr)
+	}
+
+	receiverFactory := factory.(component.ReceiverFactoryOld)
+
+	cfg, err := run.loadRuntimeReceiverConfig(receiverFactory, receiver, discoveredConfig)
+	if err != nil {
+		return nil, err
+	}
+	recvr, err := run.createRuntimeReceiver(receiverFactory, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := recvr.Start(run.ctx, run.host); err != nil {
+		return nil, fmt.Errorf("failed starting receiver %s: %v", cfg.Name(), err)
+	}
+
+	return recvr, nil
+}
+
+// shutdown the given receiver.
+func (run *receiverRunner) shutdown(rcvr component.Receiver) error {
+	return rcvr.Shutdown(run.ctx)
+}
+
+// loadRuntimeReceiverConfig loads the given receiverTemplate merged with config values
+// that may have been discovered at runtime.
+func (run *receiverRunner) loadRuntimeReceiverConfig(
+	factory component.ReceiverFactoryOld,
+	receiver receiverConfig,
+	discoveredConfig userConfigMap) (configmodels.Receiver, error) {
```

Comment:
We usually use this style in the codebase:
```suggestion
	discoveredConfig userConfigMap,
) (configmodels.Receiver, error) {
```